docs/blog/posts/pydantic.md

```diff
@@ -0,0 +1,304 @@
+---
+draft: False
+date: 2023-12-11
+slug: pydantic
+tags:
+  - pydantic
+authors:
+  - jxnl
+---
+
+# Steering Large Language Models with Pydantic
+
+In the past year, significant progress has been made in utilizing large language models. Prompt engineering, in particular, has gained attention, and new prompting techniques are being developed to guide language models toward specific tasks. While many are building chat bots, an even more exciting application is the generation of structured outputs, whether its extracting structured data, augmenting your RAG application, or even generating synthetic data.
+
+!!! question "What is Prompt Engineering?"
+
+    Prompt Engineering, also known as In-Context Prompting, is a method used to guide the behavior of LLMs without updating the model. It involves techniques to enhance the quality of outputs, formatting, reasoning, and factuality. You can learn more about it in [this post](https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/).
+
+While some have resorted to [threatening human life](https://twitter.com/goodside/status/1657396491676164096?s=20) to generate structured data, we have found that Pydantic even more effective.
+
+In this post, we will explore how we can easily validate structured outputs from language models using Pydantic and OpenAI, to write code that we can trust, then we will introduce a new library called "instructor" that simplifies this process even further while adding additional features as well.
+
+## Pydantic
+
+Unlike libraries like `dataclasses`, `Pydantic` goes a step further and allows you to define a schema for your dataclass. This schema can be used to validate data, but also to generate documentation and even to generate a JSON schema, which is perfect for our use case of generating structured data with language models!
+
+??? note "Understanding Validation"
+
+    A simple example of validation involves ensuring that a value has the correct type. For instance, let's consider a `Person` dataclass with a `name` field of type `str`. We can validate that the value is indeed a string.
+
+    ```python
+    from dataclasses import dataclass
+
+    @dataclass
+    class Person:
+        name: str
+        age: int
+
+    Person(name="Sam", age="10")
+    >>> Person(name="Sam", age="10")
+    ```
+
+    By using the `dataclass` decorator, we can pass in the values as strings without any complaints from the dataclass.
+    This would mean that we could run into issues later on if we try to use the `age` field as an `int`.
+
+    ```python
+    Person(name="Sam", age="10").age + 1
+    >>> TypeError: can only concatenate str (not "int") to str
+    ```
+
+    However, if we use `Pydantic`, we will obtain the correct type!
```

Comment:
Could add a note about how `Pydantic` validates a type and is capable of _coercion_ in order to obtain the correct type. Obviously your example is showing that, but could be helpful to make a note as well.

Reply:
We took some time to write out some extra info about what "validation" entails in `pydantic` - see this section of the docs for a reference: https://docs.pydantic.dev/latest/concepts/models/. Specifically, the "validation - a deliberate misnomer" note